api/api_debug_util.go

```diff
@@ -91,3 +116,208 @@ func (api *DebugUtilAPI) PrintBlock(ctx context.Context, blockNrOrHash rpc.Block
 	}
 	return spew.Sdump(block), nil
 }
+
+type CompactionRange struct {
+	Start    hexutil.Bytes `json:"start"`
+	Step     hexutil.Bytes `json:"step"`
+	End      hexutil.Bytes `json:"end"`
+	DB       []string      `json:"db"` // if empty, all database.
+	Interval string        `json:"interval"`
+}
+
+// CompactRange compacts over the given range, for all specified databases.
+func (r *CompactionRange) compactRange(ctx context.Context, chainDB database.DBManager) error {
+	var interval time.Duration
+	if r.Interval != "" {
+		if duration, err := time.ParseDuration(r.Interval); err != nil {
+			return err
+		} else {
+			interval = duration
+		}
+	}
+
+	fn := func(segStart, segEnd []byte) error {
+		sRange := fmt.Sprintf("0x%x-0x%x", segStart, segEnd)
+		logger.Info("Compacting database started", "db", r.DB, "range", sRange)
+		tStart := time.Now()
+		if err := r.compactSegment(ctx, chainDB, segStart, segEnd); err != nil {
+			logger.Error("Database compaction failed", "db", r.DB, "range", sRange, "err", err)
+			return err
+		}
+		tElapsed := common.PrettyDuration(time.Since(tStart))
+		logger.Info("Compacting database completed", "db", r.DB, "range", sRange, "elapsed", tElapsed)
+		time.Sleep(interval)
+		return nil
+	}
+	if err := iterateRange(r.Start, r.Step, r.End, fn); err != nil {
+		return err
+	}
+	return nil
+}
+
+// compactSegment performs the compaction for a single segment, for all specified databases.
+func (r *CompactionRange) compactSegment(ctx context.Context, chainDB database.DBManager, segStart, segEnd []byte) error {
+	if len(r.DB) == 0 {
+		return chainDB.Compact(segStart, segEnd)
+	}
+
+	for _, db := range r.DB {
+		if db, err := getDatabase(chainDB, db); err != nil {
+			return err
+		} else if compactErr := db.Compact(segStart, segEnd); compactErr != nil {
+			return compactErr
+		}
+	}
+	return nil
+}
+
+func getDatabase(dbm database.DBManager, dbName string) (database.Database, error) {
+	var db database.Database
+	switch dbName {
+	case "misc":
+		db = dbm.GetMiscDB()
+	case "header":
+		db = dbm.GetHeaderDB()
+	case "body":
+		db = dbm.GetBodyDB()
+	case "receipts":
+		db = dbm.GetReceiptsDB()
+	case "txlookup":
+		db = dbm.GetTxLookupEntryDB()
+	case "statetrie":
+		db = dbm.GetStateTrieDB()
+	default:
+		return nil, fmt.Errorf("unknown database name: %s", dbName)
+	}
+	if db == nil {
+		return nil, fmt.Errorf("cannot get database: %s", dbName)
+	}
+	return db, nil
+}
+
+func padBytes(b []byte, padLen int) []byte {
+	if len(b) >= padLen {
+		return b
+	} else {
+		return append(b, bytes.Repeat([]byte{0}, padLen-len(b))...)
+	}
+}
+
+func padBig(b *big.Int, padLen int) []byte {
+	return padBytes(b.Bytes(), padLen)
+}
+
+func padRange(start, step, end []byte) (startBig, stepBig, endBig *big.Int, padLen int, err error) {
```

Comment:
Is padding and length alignment necessary? Also, padding scheme seems confused.

Given start and end range, (start: 0x1, end: 0x200)
I thought that:
start: 0x1, end: 0x200 => start: 0x001, end: 0x200
but, it would
start: 0x1, end: 0x200 => start: 0x100, end: 0x200

If  given, start: 0x1, end: 0x100: (only changed end from 0x200 to 0x100)
start: 0x1, end: 0x100 => start: 0x100, end: 0x100 => nothing compresse?

Could you explain why padding is necessary and how it works?